packages/protocol/engine/throughputquota/mana1/throughputquota.go

```diff
@@ -0,0 +1,195 @@
+		quotaByIDCache:      shrinkingmap.New[identity.ID, int64](),
+	}, opts, func(m *ThroughputQuota) {
+		m.quotaByIDStorage.Iterate([]byte{}, func(key identity.ID, value storable.SerializableInt64) (advance bool) {
+			m.updateMana(key, int64(value))
+			return true
+		})
+
+		if totalBalanceBytes, err := m.totalBalanceStorage.Get([]byte{0}); err == nil {
+			totalBalance := new(storable.SerializableInt64)
+			if _, err := totalBalance.FromBytes(totalBalanceBytes); err != nil {
+				panic(err)
+			}
+			m.totalBalance = int64(*totalBalance)
+		}
+
+		m.engine.SubscribeConstructed(func() {
+			m.engine.Storage.Settings.SubscribeInitialized(func() {
+				m.SetLastCommittedEpoch(m.engine.Storage.Settings.LatestCommitment().Index())
+			})
+
+			m.engine.LedgerState.UnspentOutputs.Subscribe(m)
+			m.engine.LedgerState.SubscribeInitialized(func() {
+				m.engine.LedgerState.UnspentOutputs.Unsubscribe(m)
+
+				m.TriggerInitialized()
+
+				m.engine.Ledger.Events.TransactionAccepted.Attach(event.NewClosure(func(event *ledger.TransactionEvent) {
+					for _, createdOutput := range event.CreatedOutputs {
+						m.ApplyCreatedOutput(createdOutput)
+					}
+
+					for _, spentOutput := range event.SpentOutputs {
+						m.ApplySpentOutput(spentOutput)
+					}
+				}))
+
+				m.engine.Ledger.Events.TransactionOrphaned.Attach(event.NewClosure(func(event *ledger.TransactionEvent) {
+					for _, createdOutput := range event.CreatedOutputs {
+						m.ApplySpentOutput(createdOutput)
+					}
+
+					for _, spentOutput := range event.SpentOutputs {
+						m.ApplyCreatedOutput(spentOutput)
+					}
+				}))
+			})
+		})
+	})
+}
+
+// NewProvider returns a new throughput quota provider that uses mana1.
+func NewProvider(opts ...options.Option[ThroughputQuota]) engine.ModuleProvider[throughputquota.ThroughputQuota] {
+	return engine.ProvideModule(func(e *engine.Engine) throughputquota.ThroughputQuota {
+		return New(e, opts...)
+	})
+}
+
+// Balance returns the balance of the given identity.
+func (m *ThroughputQuota) Balance(id identity.ID) (mana int64, exists bool) {
+	m.quotaByIDMutex.RLock()
+	defer m.quotaByIDMutex.RUnlock()
+
+	return m.quotaByIDCache.Get(id)
+}
+
+// BalanceByIDs returns the balances of all known identities.
+func (m *ThroughputQuota) BalanceByIDs() (manaByID map[identity.ID]int64) {
+	m.quotaByIDMutex.RLock()
+	defer m.quotaByIDMutex.RUnlock()
+
+	return m.quotaByIDCache.AsMap()
+}
+
+// TotalBalance returns the total amount of throughput quota.
+func (m *ThroughputQuota) TotalBalance() (totalMana int64) {
+	m.totalBalanceMutex.RLock()
+	defer m.totalBalanceMutex.RUnlock()
+
+	return m.totalBalance
+}
+
+func (m *ThroughputQuota) ApplyCreatedOutput(output *ledger.OutputWithMetadata) (err error) {
+	if iotaBalance, exists := output.IOTABalance(); exists {
+		m.updateMana(output.AccessManaPledgeID(), int64(iotaBalance))
+
+		if !m.engine.LedgerState.UnspentOutputs.WasInitialized() {
```

Comment:
Because the total balance is the sum of all _initial_ outputs. Any other output happening after that it is just a repledge, but the total does never change anymore.